### PR TITLE
Refactor futures

### PR DESCRIPTION
**Pull request by michalmo**

### CODE REVIEW

File: asgi_rabbitmq/core.py [hunk 1]
```diff
@@ -957,16 +977,17 @@ def _schedule(self, f, *args, **kwargs):
...
+                try:
+                    result = yield deferred_from_future(
+                        self.thread.schedule(RECEIVE, channels, block=True)
+                    )
+                    defer.returnValue(result)
+                except ConnectionClosed:
+                    delay = self.thread.connection.parameters.retry_delay
+                    logger.warning(
+                        'Disconnected, retrying after %s seconds', delay)
+                    deferred_sleep(delay)
```
Comment (vshlapakov) [inline asgi_rabbitmq/core.py]: Shouldn't it be used with `yield` to work with deferred & `inlineCallbacks` properly?

File: asgi_rabbitmq/core.py [hunk 2]
```diff
@@ -26,7 +26,27 @@
...
+            except:
+                result = failure.Failure()
+            reactor.callFromThread(deferred.callback, result)
+
+        future.add_done_callback(resolve_deferred)
+        return deferred
+
+    def deferred_sleep(delay):
+        deferred = defer.Deferred()
+        reactor.callLater(delay, deferred.callback, None)
```
Comment (vshlapakov) [inline asgi_rabbitmq/core.py]: Nice trick :+1:

File: asgi_rabbitmq/core.py [hunk 3]
```diff
@@ -65,34 +83,40 @@ def __init__(self, expiry, group_expiry, get_capacity, crypter, resolve):
...
         self.are_confirmations_enabled = False
 
-    # Utilities.
-
-    def register_channel(self, method, amqp_channel):
-        """
-        Assign open AMQP channel to this protocol instance.  Call passed
-        method afterwards.  Used as AMQP channel open callback mostly.
-        """
+        self.amqp_channel = create_channel(self.apply_waiting)
```
Comment (vshlapakov) [inline asgi_rabbitmq/core.py]: My first guess was that `create_channel` is a handler function here, but I see `Protocol` [is created](https://github.com/scrapinghub/asgi_rabbitmq/blob/98a0a948b29df268a8445671759507a878724be6/asgi_rabbitmq/core.py#L773-L775) with a channel object as a first arg, could you please clarifiy it?

File: asgi_rabbitmq/core.py (hunk 1 above)
Comment (michalmo) [reply asgi_rabbitmq/core.py]: yes, my mistake

Comment (michalmo) [reply asgi_rabbitmq/core.py]: updated

File: asgi_rabbitmq/core.py (hunk 3 above)
Comment (michalmo) [reply asgi_rabbitmq/core.py]: renamed the argument to channel_factory, and used it as a keyword argument when calling the constructor to make it a bit more clear that the value passed in is a factory method (despite the unfortunate name of the method in pika)

**Pull request merged**